hoomd/mpcd/SphereGeometry.h

```diff
@@ -0,0 +1,209 @@
+            /*
+            r(t) = sqrt(x**2 + y**2 + z**2)
+            radial velocity, r_dot = (x*vel_x + y*vel_y + z*vel_z)/r(t)
+
+            define variable vr = dot(pos,vel) to check for radial motion, for computational efficiency.
+
+            To avoid division by zero error later on (line 103), we exit immediately. (no collision could have occurred
+            if the particle speed is equal to zero)
+            */
+            Scalar vr = dot(pos,vel);
+            Scalar v2 = dot(vel,vel);
+            if (r2 < R*R || vr == Scalar(0) || v2 == Scalar(0))
+               {
+               dt = Scalar(0);
+               return false;
+               }
+
+            // Find the point of contact when the particle is just leaving the surface
+            /*
+            * Calculating the time spent outside-bounds requires the knowledge of the point of contact on the boundary.
+            * The point of contact can be calculated using geometrical considerations.
+            *
+            * We know the following quantities,
+            *    1. r(t+del_t)      (the point outside the sphere)
+            *    2. v(t)            (previous velocity)
+            *
+            * Assuming 'r*' is the point of contact on the spherical shell and dx* is the distance travelled by the
+            * particle outside the boundary.
+            *
+            * r* = r(t+del_t) - dx* * v(t)/|v|
+            * and |r|**2 = R**2
+            *
+            * Solving the above equations,
+            * dx* = (r.v^) - sqrt((r.v^)**2 - r2 + R2)
+            * consequently, dt = dx* / |v|
+            */
+
+            Scalar mod_v = fast::sqrt(v2);
+            Scalar r_dot_v_hat = vr/mod_v;
+
+            const Scalar dx_ = r_dot_v_hat - fast::sqrt(r_dot_v_hat*r_dot_v_hat - r2 + R*R);
+            dt = dx_/mod_v;
+
+            // backtrack the particle for time dt to get to point of contact
+            pos.x -= vel.x*dt;
+            pos.y -= vel.y*dt;
+            pos.z -= vel.z*dt;
+
+            // update velocity according to boundary conditions
+            if (m_bc == boundary::no_slip)
+                {
+                // no-slip requires reflection of the tangential components.
+                // Radial component reflected since no penetration of the surface is necessary.
+                vel.x = -vel.x;
+                vel.y = -vel.y;
+                vel.z = -vel.z;
+                }
+            else if (m_bc == boundary::slip)
+                {
+                // tangential components of the velocity is unchanged.
+                // Radial component reflected since no penetration of the surface is necessary.
+                /*
+                 * v' = -v_perp + v_para = v - 2*v_perp
+                */
+                r2 = dot(pos,pos);
+                vr = dot(pos,vel);
+                vel.x -= 2.0*vr*pos.x/r2;
+                vel.y -= 2.0*vr*pos.y/r2;
+                vel.z -= 2.0*vr*pos.z/r2;
```

Comment:
You already know r2 is R*R, so you should use that.
```suggestion
                const Scalar3 vperp = dot(vel,pos)*pos/R;
                vel -= Scalar(2)*vperp;
```